myhoard/controller.py

```diff
@@ -265,6 +268,18 @@ def mark_backup_requested(
                     return
             self.state_manager.update_state(backup_request=new_request)
 
+    def mark_backup_preservation(self, stream_id: str, preserve_until: Optional[datetime.datetime]) -> None:
+        backup_to_preserve = self._get_backup_by_stream_id(stream_id)
+        if not backup_to_preserve:
+            raise Exception(f"Stream {stream_id} was not found in completed backups.")
+
+        with self.lock:
+            current_requests = self.state["pending_preservation_requests"]
+            current_requests[stream_id] = preserve_until.isoformat() if preserve_until else None
```

Comment:
Afair there's a gotcha with how the state update works as it compares current state against previously persisted state and if you modify the state in-place it doesn't see the change and won't persist it on disk. In-place upgrades also have more potential for introducing concurrency issues. This needs to instead be written as
```
new_requests = dict(self.state["pending_preservation_requests"])
new_requests[stream_id] = preserve_until.isoformat() if preserve_until else None
```
so that a copy of the streams is updated instead of the original one.

Reply:
Updated: https://github.com/Aiven-Open/myhoard/blob/02605fbb08f5d6cfa3ff34aa34e852d81b4d2c95/myhoard/controller.py#L278-L281